src/test/java/co/rsk/federate/btcreleaseclient/cache/PegoutSignedCacheImplTest.java

```diff
@@ -0,0 +1,113 @@
+
+class PegoutSignedCacheImplTest {
+
+  private static final Duration DEFAULT_TTL = Duration.ofMinutes(30);
+  private static final Keccak256 PEGOUT_CREATION_RSK_HASH = TestUtils.createHash(1);
+
+  private final Map<Keccak256, Instant> cache = new ConcurrentHashMap<>();
+  private final PegoutSignedCache pegoutSignedCache = PegoutSignedCacheFactory.from(DEFAULT_TTL);
+
+  @BeforeEach
+  void setUp() throws Exception {
+    Field field = pegoutSignedCache.getClass().getDeclaredField("cache");
+    field.setAccessible(true);
+    field.set(pegoutSignedCache, cache);
+  }
+
+  @Test
+  void hasAlreadyBeenSigned_shouldReturnFalse_whenPegoutCreationRskTxHashIsNull() {
+    Keccak256 pegoutCreationRskTxHash = null;
+
+    boolean result = pegoutSignedCache.hasAlreadyBeenSigned(pegoutCreationRskTxHash);
+
+    assertFalse(result);
+  }
+
+  @Test
+  void hasAlreadyBeenSigned_shouldReturnFalse_whenPegoutCreationRskTxHashIsNotCached() {
+    boolean result = pegoutSignedCache.hasAlreadyBeenSigned(PEGOUT_CREATION_RSK_HASH);
+
+    assertFalse(result);
+  }
+
+  @Test
+  void hasAlreadyBeenSigned_shouldReturnFalse_whenCacheContainsExpiredTimestamp() {
+    Instant currentTimestamp = Instant.now();
+    Instant timestamp = currentTimestamp.minusMillis(Duration.ofMinutes(60).toMillis());
+    cache.put(PEGOUT_CREATION_RSK_HASH, timestamp);
+
+    boolean result = pegoutSignedCache.hasAlreadyBeenSigned(PEGOUT_CREATION_RSK_HASH);
+
+    assertFalse(result);
+  }
+
+  @Test
+  void hasAlreadyBeenSigned_shouldReturnTrue_whenCacheContainsNotExpiredTimestamp() {
+    Instant currentTimestamp = Instant.now();
+    Instant timestamp = currentTimestamp.minusMillis(Duration.ofMinutes(10).toMillis());
+    cache.put(PEGOUT_CREATION_RSK_HASH, timestamp);
+
+    boolean result = pegoutSignedCache.hasAlreadyBeenSigned(PEGOUT_CREATION_RSK_HASH);
+
+    assertTrue(result);
+  }
+
+  @Test
+  void putIfAbsent_shouldNotPutInCache_whenPegoutCreationRskTxHashIsNull() {
+    Keccak256 pegoutCreationRskTxHash = null;
+
+    pegoutSignedCache.putIfAbsent(pegoutCreationRskTxHash);
+
+    assertFalse(cache.containsKey(PEGOUT_CREATION_RSK_HASH));
+  }
+
+  @Test
+  void putIfAbsent_shouldPutInCache_whenPegoutCreationRskTxHashIsNotNull() {
+    pegoutSignedCache.putIfAbsent(PEGOUT_CREATION_RSK_HASH);
+
+    assertTrue(cache.containsKey(PEGOUT_CREATION_RSK_HASH));
+  }
+
+  @Test
+  void putIfAbsent_shouldPutInCacheBoth_whenPegoutCreationRskTxHashAreNotSame() {
+    // first insert
+    pegoutSignedCache.putIfAbsent(PEGOUT_CREATION_RSK_HASH);
+    Instant pegoutCreationRskTxHashTimestamp = cache.get(PEGOUT_CREATION_RSK_HASH);
+    // second insert
+    Keccak256 otherPegoutCreationRskTxHash = TestUtils.createHash(2);
+    pegoutSignedCache.putIfAbsent(otherPegoutCreationRskTxHash);
+    Instant otherPegoutCreationRskTxHashTimestamp = cache.get(otherPegoutCreationRskTxHash);
+
+    assertNotSame(pegoutCreationRskTxHashTimestamp, otherPegoutCreationRskTxHashTimestamp);
```

Comment:
to stick with the test name i would first assert that both rsk hashes are in the cache.
I know that that's implied when getting the timestamp using their keys, but I think when you read the test name you are expecting this to happen... checking the timestamps are different is an addition to the test, not its "main purpose".
Wdyt?

```
    // first insert
    pegoutSignedCache.putIfAbsent(PEGOUT_CREATION_RSK_HASH);
    
    // second insert
    Keccak256 otherPegoutCreationRskTxHash = TestUtils.createHash(2);
    pegoutSignedCache.putIfAbsent(otherPegoutCreationRskTxHash);

   assertTrue(cache.containsKey(PEGOUT_CREATION_RSK_HASH));
   assertTrue(cache.containsKey(otherPegoutCreationRskTxHash));   


    // also check that they have different timestamps
    Instant pegoutCreationRskTxHashTimestamp = cache.get(PEGOUT_CREATION_RSK_HASH);
    Instant otherPegoutCreationRskTxHashTimestamp = cache.get(otherPegoutCreationRskTxHash);

    assertNotSame(pegoutCreationRskTxHashTimestamp, otherPegoutCreationRskTxHashTimestamp);
```

Reply:
Not sure, checking if the timestamps are not the same (i.e. don't hold the same reference) is the main purpose of the test. The other asserts you mention are already covered in their own test even though as you mention it is implied and could be placed as well. Ideally we should test one thing only per test, "ideally".

Reply:
i agree with what you are saying, but when i read the test name i expect to check if both values are in the cache

but leave it that way if you think is clear enough! no worries